Changes for async API (non-blocking reqwest)

leodutra opened the pull request:
Implements `async send()` + `async fetch_resource()` and implements changes for the new usage (+examples). 

This change is required to enable the API to be used in async scopes using async runtime, like actix-web 2.0. Otherwise, a "trying to block into async thread" [sic] error occurs.

The examples use `futures` crate to keep the code as close to the last version as possible.

I was able to test into my own private chatbot, it's working as expected.

This should probably become a 4.x version.

biomunky commented:
Hi, can you bump the version. Let’s get this in.

biomunky commented:
When I run the examples I'm getting a panic

Running `target/debug/examples/get_sources`
thread 'main' panicked at 'not currently running on the Tokio runtime.', /Users/dan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.16/src/runtime/handle.rs:69:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

biomunky commented:
The panic can be solved by doing the following.

Add tokio to Cargo.toml

    tokio = { version = "0.2", features = ["full"] }

Modify the examples to use the tokio main macro. Using full as this is recommended in the docs when starting out.

    #[tokio::main]
    async fn main() { ... }

While the async stuff is cool, I'm not sure I'm comfortable with throwing the synchronous client away. Perhaps a refactor is in order so that we can support async and sync clients?

leodutra commented:
Sounds good. It's late my time, and sorry for the delay, but I'll do it ASAP.

biomunky commented:
> Sounds good. It's late my time, and sorry for the delay, but I'll do it ASAP.

No rush, this is supposed to be fun rather than work.

leodutra commented:
Hey @biomunky. Sorry, I totally missed this PR.
I kept the sync API, added some `_sync` suffix for the 2 sync/ async methods we have,
fixed the examples and repeated them for sync versions.

biomunky commented:
cool, likewise missed :) will look tomorrow.

biomunky commented on src/api.rs:
```diff
@@ -128,7 +146,32 @@ impl Client {
 
-    fn fetch_resource(url: &str, api_key: &str) -> Result<String, NewsApiError> {
+    async fn fetch_resource(url: &str, api_key: &str) -> Result<String, NewsApiError> {
+        static CLIENT_USER_AGENT: &str = concat!(
+            "rust-",
+            env!("CARGO_PKG_NAME"),
+            "/",
+            env!("CARGO_PKG_VERSION"),
+        );
```
This duplicated bit of code can be factored out into a function

```
    fn create_user_agent() -> String {
      concat!(
          "rust-",
          env!("CARGO_PKG_NAME"),
          "/",
          env!("CARGO_PKG_VERSION"),
      ).to_owned()
    }
```

leodutra replied:
This is an old code. For some reason, it was marked as changed. :thinking: 
But I can change this.

leodutra replied:
This is some existing code. For some reason, it was marked as changed. thinking
But I can change this.

biomunky commented on src/api.rs:
```diff
@@ -62,7 +62,7 @@ impl Client {
 
     /// Send the constructed URL to the newsapi server
-    pub fn send<T>(&self) -> Result<T, NewsApiError>
+    pub async fn send<T>(&self) -> Result<T, NewsApiError>
```
Maybe we should make the methods explicit, if we're changing send to be async why not go with send_sync and send_async and do the same for the other methods?

leodutra replied:
Personally, I think async is more interesting for the future of the lib since it is an API wrapper.
Async is a big jump and probably a good reason for a breaking change since in the future most users will be expecting async.

biomunky commented:
I'm going to merge and then make some more changes before releasing a new version.
